### ISSUE TITLE
Unable to obtain lock file access on '/tmp/NuGetScratch/lock/...' since .NET 7

### ISSUE THREAD

**Issue by kartheekp-ms** [NuGet/Home#12299]
### NuGet Product Used

dotnet.exe

### Product Version

.NET 7 SDK

### Impact

I'm unable to use this version

### Repro Steps & Context

@socolin created https://github.com/dotnet/sdk/issues/29394. I moved it to NuGet repository as the issue is related to NuGet functionality.

### Describe the bug

I have 2 users on my computer (Linux, Ubuntu 20.04) one for myself and one used by an azure devops agent that build the solution. 

When I'm working on the solution with my user, the command `dotnet restore` is failing on the build server

This was working fine until I installed .NET 7 sdk

```
Unable to obtain lock file access on '/tmp/NuGetScratch/lock/9d44d83c2af3cfd899a033f524442ea238044935' for operations on '/home/azure-agent/agent1/_work/1/Nuget/tempNuGet_13529.config'. This may mean that a different user or administrator is holding this lock and that this process does not have permission to access it. If no other process is currently performing an operation on this file it may mean that an earlier NuGet process crashed and left an inaccessible lock file, in this case removing the file '/tmp/NuGetScratch/lock/9d44d83c2af3cfd899a033f524442ea238044935' will allow NuGet to continue.
```

### To Reproduce

2 users on linux, trying to restore a similar project  at the same time.

### Further technical details

```
.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.100/

Host:
  Version:      7.0.0
  Architecture: x64
  Commit:       d099f075e4

.NET SDKs installed:
  3.1.425 [/usr/share/dotnet/sdk]
  5.0.408 [/usr/share/dotnet/sdk]
  6.0.403 [/usr/share/dotnet/sdk]
  7.0.100 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.31 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.31 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by kartheekp-ms** [comment]
@socolin - Is this an intermittent issue or you were able to reproduce this error every time when multiple users run restore on the machine at the same time? I am trying to understand if you were able to restore the solution/project successfully on .NET6 for this scenario.

> Unable to obtain lock file access on '/tmp/NuGetScratch/lock/9d44d83c2af3cfd899a033f524442ea238044935'

I assume the lock file (in this case "/tmp/NuGetScratch/lock/9d44d83c2af3cfd899a033f524442ea238044935") is owned by your account which the ci-agent can't access. Do these different users share the global packages folder by any chance?

**Comment by Socolin** [comment]
> I assume the lock file (in this case "/tmp/NuGetScratch/lock/9d44d83c2af3cfd899a033f524442ea238044935") is owned by your account which the ci-agent can't access. Do these different users share the global packages folder by any chance?

No, when it happen it does not stop, I have to remove those file manually

This is owned by my user, and it fail with the CI. I'm using Rider and when it build it left some dotnet process running to be reused later (/nodeReuse:true)

The CI il building the same project (but in a different folder with a different user) and it fail when those files exists

**Comment by Nefcanto** [comment]
I also see this bug when I want to debug (F5) my application inside the container. I'm using Dev Containers extension for attaching to containers. It was working just fine in the past months, but recently I got these errors.

Any idea how should we solve this?

If I run `sudo dotnet build` it works. But I can't make VS Code to use `sudo` when I press F5. And it keeps downloading the dependencies.

**Comment by zivkan** [comment]
As a workaround, you can set the `NUGET_SCRATCH` environment variable to different directories per account, to make NuGet use different tmp paths for different user accounts. A product fix is on the way.

**Comment by Nefcanto** [comment]
@zivkan, I did it but it did not work. In my `docker-compose.yml` file under the `environment` key I added `- NUGET_SCRATCH=/Temp/NuGet` and I still see that `vscode` user can not take the lock. Any ideas?

**Comment by zivkan** [comment]
when you check the `lock/` directory that the error complains about, are different files owned by different users? If so, you need to delete them one time, then as long as you're consistent in using the environment variable to force different NuGetScratch directories, there shouldn't be ongoing issues.

If you're using .NET 7 or higher, then you can set the env var `NUGET_ConcurrencyUtils_DeleteOnClose` to value `1`. That will tell NuGet to use the BCL's DeleteOnClose flag on the file, but .NET Core 1.0 to .NET 6 had a bug where delete on close and `FileShare.None` don't work together.

**Comment by Nefcanto** [comment]
@zivkan here is an image of the VS Code terminal (VS Code is attached to the container already):

![Screenshot from 2023-02-10 10-16-27](https://user-images.githubusercontent.com/12782347/218022119-cb2a0bb9-af41-4bd6-9db9-2a1baf4528ad.png)

As you can see, both `NUGET_SCRATCH` and `NUGET_ConcurrencyUtils_DeleteOnClose` are set.

And as you can see, I delete `/Temp/NuGet` and change the ownership of the `/Temp` directory altogether and give it to the `vscode` user.

Now everything is clean and I run `dotnet build`, and this is what I get:

![Screenshot from 2023-02-10 10-20-16](https://user-images.githubusercontent.com/12782347/218022502-63228ff9-49ce-4195-8801-fda4800de11f.png)

As you can see, it tries to install packages (though it times out, packages are already cached and it should not)

And at the end I see this error:

![Screenshot from 2023-02-10 10-21-38](https://user-images.githubusercontent.com/12782347/218022717-ee2f8723-ae1e-4588-9ff0-4c60388c2c08.png)

In this case, the error is something new, thus I ran `sudo rm -rf obj` and `sudo rm -rf bin` and:

![Screenshot from 2023-02-10 10-22-43](https://user-images.githubusercontent.com/12782347/218023054-0d99236d-2bc0-4fd8-b4b3-7d4a4734036e.png)

But I see this message:

> Unable to use package assets cache due to I/O error. This can occur when the same project is built more than once in parallel. Performance may be degraded, but the build result will not be impacted.

But if I use `sudo dotnet build` then everything is OK:

![Screenshot from 2023-02-10 10-24-21](https://user-images.githubusercontent.com/12782347/218023244-4389e7fb-b9a0-4ae1-a097-50b742f514b8.png)

The problem is that I can't (or I don't know how to) bind VS Code's F5 (or Run => Start Debugging) with all of these bash commands. And in this case, it did not give me errors related to locks, but many times it complained about locks too.

**Comment by Nefcanto** [comment]
I wish we could run everything as `root` user and not tolerate all of these pains. We take the responsibility of developing with the root user. Microsoft please hear us. At least give us this option to choose to have less security, but also less pain of dealing with your decisions.

**Comment by Nefcanto** [comment]
This time I performed this operations before doing the build:

```
sudo rm -rf obj
sudo rm -rf bin
sudo rm -rf /Temp/NuGet
sudo chown -R vscode:vscode /Temp
```

Then I ran `dotnet build` without sudo inside the VS Code terminal, and it got built. 

But it still downloaded all packages and took 3 minutes to build. while when I use `sudo dotnet build` it would restore the project in 500ms or less.

But I'm still stuck on two things. I can't automate that code into the VS Code F5, and I can't make it reuse the existing NuGet cache.
